keras_hub/src/models/deepseek_r1/deepseek_layers.py

```diff
@@ -0,0 +1,791 @@
+            dim (int): Dimensionality of the embedding space.
+            base (float): Base value for the exponential computation.
+            max_seq_len (int): Maximum sequence length.
+
+        Returns:
+            Tuple[int, int]: The range of correction dimensions (low, high), clamped to valid indices.
+        """
+        low = ops.floor(find_correction_dim(low_rot, dim, base, max_seq_len))
+        high = ops.ceil(find_correction_dim(high_rot, dim, base, max_seq_len))
+        return max(low, 0), min(high, dim - 1)
+
+    def linear_ramp_factor(min, max, dim):
+        """
+        Computes a linear ramp function used to smooth values between a minimum and maximum range.
+
+        Args:
+            min (float): Minimum value for the ramp function.
+            max (float): Maximum value for the ramp function.
+            dim (int): Dimensionality of the ramp tensor.
+
+        Returns:
+            torch.Tensor: A tensor of shape (dim,) with values linearly interpolated between 0 and 1,
+                clamped to the range [0, 1].
+        """
+        if min == max:
+            max += 0.001
+        linear_func = (ops.arange(dim) - min) / (max - min)
+        ramp_func = ops.clip(linear_func, 0, 1)
+        return ramp_func
+
+    freqs = 1.0 / (base ** (ops.arange(0, dim, 2) / dim))
+    if seqlen > original_seq_len:
+        low, high = find_correction_range(
+            beta_fast, beta_slow, dim, base, original_seq_len
+        )
+        smooth = 1 - linear_ramp_factor(low, high, dim // 2)
+        freqs = freqs / factor * (1 - smooth) + freqs * smooth
+
+    t = ops.arange(seqlen)
+    freqs = ops.outer(t, freqs)
+    freqs_cis = ops.polar(ops.ones_like(freqs), freqs)
+    return freqs_cis
+
+
+def _view_as_complex(tensor):
+    """
+    Tensor shaped (M, N, 2) ->
+    """
+    real_part = tensor[..., 0]
+    imag_part = tensor[..., 1]
+    return real_part + 1j * imag_part
+
+
+def _view_as_real(x):
+    # Assuming x is a complex tensor, we extract the real and imaginary parts
+    real_part = ops.real(x)
+    imag_part = ops.imag(x)
+
+    # Stack the real and imaginary parts along the last axis
+    return ops.stack((real_part, imag_part), axis=-1)
+
+
+def apply_rotary_emb(x, freqs_cis):
+    """
+    Applies rotary positional embeddings to the input tensor.
+
+    Args:
+        x (torch.Tensor): Input tensor with positional
+            embeddings to be applied.
+        freqs_cis (torch.Tensor): Precomputed complex exponential values
+          for positional embeddings.
+
+    Returns:
+        torch.Tensor: Tensor with rotary embeddings applied.
+    """
+    dtype = x.dtype
+    x = _view_as_complex(ops.reshape(x, [*x.shape[:-1], -1, 2]))
+    freqs_cis = ops.reshape(freqs_cis, [1, x.shape[1], 1, x.shape[-1]])
+    y = _view_as_real(x * freqs_cis)
+    # flatten on axis 3
+    y = ops.reshape(y, x.shape[:3] + (-1,))
+    return y
+
+
+class MLA(layers.Layer):
+    """
+    Multi-Headed Attention Layer (MLA).
```

Comment:
change this to --> """Multi-Headed Attention Layer (MLA).

 Basically, """ should be followed by string in the same line